src/main/java/org/cloudfoundry/promregator/rewrite/CFAllLabelsMetricFamilySamplesEnricher.java

```diff
@@ -26,40 +27,56 @@
 	public static final String LABELNAME_INSTANCE_NUMBER = "cf_instance_number";
 	
 	private static String[] labelNames = new String[] { LABELNAME_ORGNAME, LABELNAME_SPACENAME, LABELNAME_APPNAME, LABELNAME_INSTANCEID, LABELNAME_INSTANCE_NUMBER };
-	private Map<String, String> labelsAndValuesToAdd = new LinkedHashMap<>();
 	public static String[] getEnrichingLabelNames() {
 		return labelNames.clone();
 	}
 
+	private String orgName;
+	private String spaceName;
+	private String appName;
+	private String instanceId;
+
 	public CFAllLabelsMetricFamilySamplesEnricher(String orgName, String spaceName, String appName, String instanceId) {
-		this.labelsAndValuesToAdd.put(LABELNAME_ORGNAME, orgName);
-		this.labelsAndValuesToAdd.put(LABELNAME_SPACENAME, spaceName);
-		this.labelsAndValuesToAdd.put(LABELNAME_APPNAME, appName);
-		this.labelsAndValuesToAdd.put(LABELNAME_INSTANCEID, instanceId);
-		this.labelsAndValuesToAdd.put(LABELNAME_INSTANCE_NUMBER, getInstanceFromInstanceId(instanceId));
+		this.orgName = orgName;
+		this.spaceName = spaceName;
+		this.appName = appName;
+		this.instanceId = instanceId;
 	}
 	
 	@Override
 	protected List<String> getEnrichedLabelNames(List<String> original) {
 		List<String> clone = new LinkedList<>(original);
-		Collections.addAll(clone, labelsAndValuesToAdd.keySet().toArray(new String[0]));
+		for(String label: labelNames) {
+			clone.remove(label);
+			//TODO: log that label was removed
+		}
+		Collections.addAll(clone, labelNames);
 		return clone;
 	}
 	
 	@Override
 	public List<String> getEnrichedLabelValues(List<String> original) {
 		List<String> clone = new LinkedList<>(original);
-		Collections.addAll(clone, labelsAndValuesToAdd.values().toArray(new String[0]));
+		clone.add(orgName);
+		clone.add(spaceName);
+		clone.add(appName);
+		clone.add(instanceId);
+		clone.add(getInstanceFromInstanceId(instanceId));
 		return clone;
 	}
 
 	@Override
-	protected void removeDuplicateLabels(List<String> original){
-		for(String label: original) {
-			labelsAndValuesToAdd.remove(label);
+	protected List<String> getEnrichedLabelValues(List<String> originalLabelNames, List<String> originalLabelValues) {
+		List<String> clone = new LinkedList<>(originalLabelValues);
+		for(String label: labelNames) {
+			if(originalLabelNames.contains(label)){
```

Comment:
This logic here is behaving bad on performance. Note that this is important here, as we need to enrich possibly hundreds (if not thousands) of metrics within a short timeframe.
Additionally `originalLabelNames` can also be a large in size. 

Consider using a precalculated `HashSet<String>` here. That could improve performance significantly.

Reply:
Yea that is a good call!